### PR TITLE
core: add FlatSymbolRefAttr

### PR DESCRIPTION
**Pull request by luisacicolini**

### CODE REVIEW

File: xdsl/dialects/builtin.py [hunk 1]
```diff
@@ -218,6 +220,22 @@ def string_value(self):
         return root
 
 
+class EmptyArrayAttrConstraint(AttrConstraint):
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:223]: Could you please add some documentation strings for this class, and the two values below?

File: xdsl/dialects/builtin.py [hunk 2]
```diff
@@ -232,8 +236,10 @@ def verify(self, attr: Attribute, constraint_vars: dict[str, Attribute]) -> None
 FlatSymbolRefAttrConstraint = ParamAttrConstraint(
     SymbolRefAttr, [AnyAttr(), EmptyArrayAttrConstraint()]
 )
+"""Constrain SymbolRef to be FlatSymbolRef"""
 
 FlatSymbolRefAttr = Annotated[SymbolRefAttr, FlatSymbolRefAttrConstraint]
+"""SymbolRef constrained to Flat"""
```
Comment (superlopuh) [inline xdsl/dialects/builtin.py:242]:
```suggestion
"""SymbolRef constrained to have an empty `nested_references` property."""
```

**Review by superlopuh — APPROVED** [review]
Thank you! Can you please expand a bit on one of the documentation strings? Otherwise good to go.

**Review by PapyChacal — APPROVED** [review]

**Review by compor — APPROVED** [review]

File: tests/filecheck/dialects/func/func_invalid.mlir [hunk 3]
```diff
@@ -31,3 +31,9 @@ builtin.module {
 }
 
 // CHECK: Expected all arguments to be named or all arguments to be unnamed.
+
+// -----
+
+"func.call"() { "callee" = @call::@invalid } : () -> ()
+
+// CHECK:  Operation does not verify: expected empty array, but got ["invalid"]
```
Comment (webmiche) [inline tests/filecheck/dialects/func/func_invalid.mlir:39]: This does not feel like an actionable error message. What is an array in this context? Can we somehow make this understandable?

Comment (math-fehr) [reply tests/filecheck/dialects/func/func_invalid.mlir:39]: We have an idea, but that's going to be for another PR.
Essentially we'll add a new constraint that will catch an error, and return it.
For now at least, this error is better than the previous one ;)

Comment (superlopuh) [reply tests/filecheck/dialects/func/func_invalid.mlir:39]: #1579

**Pull request merged**